jobrunner/lib/database.py

```diff
@@ -168,13 +172,28 @@ def get_connection(filename=None):
     # Looks icky but is documented `threading.local` usage
     cache = CONNECTION_CACHE.__dict__
     if filename not in cache:
-        conn = sqlite3.connect(filename)
+        try:
+            conn = sqlite3.connect(filename)
+        except sqlite3.OperationalError:
+            raise CannotOpenReadonlyDBFile(filename)
         # Enable autocommit so changes made outside of a transaction still get
         # persisted to disk. We can use explicit transactions when we need
         # atomicity.
         conn.isolation_level = None
         # Support dict-like access to rows
         conn.row_factory = sqlite3.Row
+        # Add any custom initial setup
+        if initial:
+            log.info("applying initial DB connection setup")
+            try:
+                conn.execute("BEGIN")
+                for statement in initial:
+                    conn.execute(statement)
+            except Exception:
+                conn.rollback()
+                raise
+            else:
+                conn.commit()
```

Comment:
I don't quite understand how this works. This is going to run the `initial` statements on every connection which are going to error once the tables exist. But I can't see what catches these errors.

Reply:
It only runs if the connection is not already in the cache.

And the DDL is explicitly idempotent, by have IF NOT EXISTS.